tools/depcheck/README.md

```diff
@@ -0,0 +1,288 @@
+  - Would those parts of our codebase work with the version of package `A` that is brought
+  in by `kubernetes` (or one of its dependencies)?
+  - If not, can we break our dependency on package `A`?
+ 
+Both the `analyze` and `trace` commands included in this tool are meant to aid in 
+answering the above questions.
+
+<a name="started"></a>
+## Getting started
+
+<a name="compiling"></a>
+### Compiling
+
+To compile the `depcheck` binary, simply use `go build` and point to the tool's entrypoint:
+
+```bash
+$ go build /path/to/openshift/origin/tools/depcheck/depcheck.go
+```
+
+After this step, you should have a `depcheck` binary in your current working directory.
+
+
+<a name="running"></a>
+### Running
+
+The `depcheck` tool can be used to obtain `dot` output for a given repository.
+This output may be fed into tools like [dot (1)](https://www.systutorials.com/docs/linux/man/1-dot/)
+or [d3-graphviz](https://github.com/magjac/d3-graphviz) in order to obtain a visual dependency graph.
+
+Example:
+
+Invoke the `trace` command using `openshift`-specific defaults, generating DOT output which can
+be used to render a graph of the repository:
+
+```bash
+$ depcheck trace --root=github.com/openshift/origin --entry=cmd/... --openshift > graph.dot
+```
+
+*Notes*
+
+- The `--root` flag contains the fully-qualified Go import-path of the target repository.
+
+Additionally, this tool can be used to obtain an analysis report of your repository's vendored packages
+against those of a vendored dependency. This output lists which packages you and them have in common, as
+well as which ones are unique to either of you.
+
+Example:
+
+Invoke the `analyze` command using `openshift`-specific defaults, comparing dependencies against those of `kubernetes`:
+
+```bash
+$ depcheck analyze --root=github.com/openshift/origin --entry=cmd/... --entry=pkg/... --entry=tools/... --entry=test/... --openshift --dep=github.com/openshift/origin/vendor/k8s.io/kubernetes
+```
+
+*Notes*
+
+- The ellipsis (`...`) after each `--entry` flag value indicate that you wish to recurse through that directory and include those packages.
+  - It is possible to omit the ellipsis if you wish to do a "shallow" analysis of only that directory, as long as the directory path provided contains at least a single Go file under it.
+- Directory names provided via `--entry` may be specified relative to the value of `--root`, or as fully qualified Go import-paths.
+
+
+The `depcheck` tool can further be used to `pin` dependencies from a `glide.lock` file onto a `glide.yaml` file.
+*Pinning* these dependencies means that you are taking all dependencies listed in a lock file that are not currently present
+in the given `glide.yaml` file, and generating a new `glide.yaml` file containing this new set of dependencies, each one pinned to
+the commit hash specified in the lock file.
+
+Example:
+
+```bash
+$ depcheck pin --glide.yaml=glide.yaml --glide.lock=glide.lock > glide.yaml
```

Comment:
Change the final `glide.yaml` to `newglide.yaml`, if you invoked above command you're get an empty glide.yaml file error.